src/Framework/Analyzer/src/WebApplicationBuilder/WebApplicationBuilderAnalyzer.cs

```diff
@@ -0,0 +1,137 @@
+        DiagnosticDescriptors.DoNotUseConfigureWithConfigureWebHostBuilder,
+        DiagnosticDescriptors.DoNotUseUseStartupWithConfigureWebHostBuilder,
+    });
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationStartAction(compilationStartAnalysisContext =>
+        {
+            var compilation = compilationStartAnalysisContext.Compilation;
+            if (!WellKnownTypes.TryCreate(compilation, out var wellKnownTypes))
+            {
+                Debug.Fail("One or more types could not be found. This usually means you are bad at spelling C# type names.");
+                return;
+            }
+
+            compilationStartAnalysisContext.RegisterOperationAction(operationAnalysisContext =>
+            {
+                var invocation = (IInvocationOperation)operationAnalysisContext.Operation;
+                var targetMethod = invocation.TargetMethod;
+
+                DisallowConfigureHostBuilderConfigureWebHost(
+                    operationAnalysisContext,
+                    wellKnownTypes,
+                    invocation,
+                    targetMethod);
+
+                DisallowConfigureWebHostBuilderConfigure(
+                    operationAnalysisContext,
+                    wellKnownTypes,
+                    invocation,
+                    targetMethod);
+
+                DisallowConfigureWebHostBuilderUseStartup(
+                    operationAnalysisContext,
+                    wellKnownTypes,
+                    invocation,
+                    targetMethod);
+
+            }, OperationKind.Invocation);
+        });
+    }
+
+    private static bool IsDisallowedMethod(
+        in OperationAnalysisContext context,
+        IInvocationOperation invocation,
+        IMethodSymbol methodSymbol,
+        INamedTypeSymbol disallowedReceiverType,
+        string disallowedMethodName,
+        params INamedTypeSymbol[] disallowedMethodTypes)
+    {
+        if (!IsDisallowedMethod(methodSymbol, disallowedMethodName, disallowedMethodTypes))
+        {
+            return false;
+        }
+
+        var receiverType = GetReceiverType(invocation, context.CancellationToken);
+
+        if (!SymbolEqualityComparer.Default.Equals(receiverType, disallowedReceiverType))
+        {
+            return false;
+        }
+
+        return true;
+
+        static bool IsDisallowedMethod(
+            IMethodSymbol methodSymbol,
+            string disallowedMethodName,
+            INamedTypeSymbol[] disallowedMethodTypes)
+        {
+            if (!string.Equals(methodSymbol?.Name, disallowedMethodName, StringComparison.Ordinal))
+            {
+                return false;
+            }
+
+            return disallowedMethodTypes.Any(type => SymbolEqualityComparer.Default.Equals(type, methodSymbol.ContainingType));
```

Comment:
Could we use a regular for loop? We should treat all of the analyzers as if they were running in the hot path all the time.